src/main/java/com/tabnine/DependencyContainer.java

```diff
@@ -2,10 +2,33 @@
 
 import com.tabnine.binary.BinaryFacade;
 import com.tabnine.binary.BinaryRun;
+import com.tabnine.binary.TabNineGateway;
 import com.tabnine.binary.fetch.*;
+import com.tabnine.selections.TabNineLookupListener;
 import org.jetbrains.annotations.NotNull;
 
 public class DependencyContainer {
+    private static TabNineGateway GATEWAY_INSTANCE = null;
+    private static TabNineLookupListener LOOKUP_LISTENER_INSTANCE = null;
+
+    public static synchronized TabNineGateway singletonOfTabNineGateway() {
+        if (GATEWAY_INSTANCE == null) {
+            GATEWAY_INSTANCE = new TabNineGateway();
+
+            GATEWAY_INSTANCE.init();
+        }
+
+        return GATEWAY_INSTANCE;
+    }
+
+    public static synchronized TabNineLookupListener singletonOfTabNineLookupListener() {
+        if (LOOKUP_LISTENER_INSTANCE == null) {
+            LOOKUP_LISTENER_INSTANCE = new TabNineLookupListener();
```

Comment:
why is it singleton? what state does it manages?

Reply:
Well they way IntelliJ does selections is weird, as it requires me to register the selection listener upon a completion suggestion (we did the same in Codota & I searched the web for how to do it), so to make sure that I won't register/initiate `TabNineLookupListener` multiple times I make it a singleton and removes it from the selection listener list before adding it again. At:
https://github.com/codota/tabnine-intellij/pull/53/files#diff-04d46f64b334bd2109e21f7780c6176fR133

Reply:
👍